droidlet/perception/robot/handlers/label_propagate.py

```diff
@@ -76,41 +145,56 @@ def __call__(
+        cur_pose,
         cur_depth,
     ):
         """
         1. Gets point cloud for the source image
-        2. Transpose the point cloud based on robot location (base_pose)
+        2. Transpose the point cloud based on robot location (cur_pose)
         3. Project the point cloud back into the image frame. The corresponding semantic label for each point from the src_label becomes
         the new semantic label in the current frame.
         Args:
             src_img (np.ndarray): source image to propagte from
             src_depth (np.ndarray): source depth to propagte from
             src_label (np.ndarray): source semantic map to propagte from
             src_pose (np.ndarray): (x,y,theta) of the source image
-            base_pose (np.ndarray): (x,y,theta) of current image
+            cur_pose (np.ndarray): (x,y,theta) of current image
             cur_depth (np.ndarray): current depth
         """
 
         height, width, _ = src_img.shape
         uv_one_in_cam, intrinsic_mat, rot, trans = compute_uvone(height, width)
 
-        pts_in_world = convert_depth_to_pcd(src_depth, src_pose, uv_one_in_cam, rot, trans)
+        src_pts_in_world = convert_depth_to_pcd(src_depth, src_pose, uv_one_in_cam, rot, trans)
 
-        # TODO: can use cur_pts_in_world for filtering. Not needed for baseline.
-        # cur_pts_in_world = convert_depth_to_pcd(cur_depth, base_pose, uv_one_in_cam, rot, trans)
+        # visualize using o3d
+        # visualize_pcd(src_pts_in_world)
 
         # convert pts_in_world to current base
-        pts_in_cur_base = transform_pose(pts_in_world, (-base_pose[0], -base_pose[1], 0))
-        pts_in_cur_base = transform_pose(pts_in_cur_base, (0.0, 0.0, -base_pose[2]))
+        src_pts_in_cur_base = transform_pose(src_pts_in_world, (-cur_pose[0], -cur_pose[1], 0))
+        src_pts_in_cur_base = transform_pose(src_pts_in_cur_base, (0.0, 0.0, -cur_pose[2]))
 
         # conver point from current base to current camera frame
-        pts_in_cur_cam = pts_in_cur_base - trans.reshape(-1)
-        pts_in_cur_cam = np.dot(pts_in_cur_cam, rot)
+        src_pts_in_cur_cam = src_pts_in_cur_base - trans.reshape(-1)
+        src_pts_in_cur_cam = np.dot(src_pts_in_cur_cam, rot)
 
-        # conver pts in current camera frame into 2D pix values
-        pts_in_cur_img = np.matmul(intrinsic_mat, pts_in_cur_cam.T).T
-        pts_in_cur_img /= pts_in_cur_img[:, 2].reshape([-1, 1])
+        # Get Valid Z
+        valid_z = src_pts_in_cur_cam[:, 2] > 0
+
+        # Filter based on current depth.
+        cur_depth = (cur_depth.astype(np.float32) / 1000.0).reshape(-1)
+        cur_pts_in_cur_cam = np.multiply(uv_one_in_cam, cur_depth).T
 
-        return get_annot(height, width, pts_in_cur_img, src_label)
+        # conver pts in current camera frame into 2D pix values
+        src_pts_in_cur_img = np.matmul(intrinsic_mat, src_pts_in_cur_cam.T).T
+        src_pts_in_cur_img /= src_pts_in_cur_img[:, 2].reshape([-1, 1])
```

Comment:
i think this is a bug. you shouldn't divide `src_pts_in_cur_img[:, 2] / src_pts_in_cur_img[:, 2]` right?
You should only do: `src_pts_in_cur_img[:, 0:2] / src_pts_in_cur_img[:, 2]`

Reply:
`src_pts_in_cur_img[:, 2]` needs to be reshaped for the division. So this does that by `src_pts_in_cur_img[:, 2].reshape([-1, 1])` which is essentially a `num_row * 1` array consisting of the last column of `src_pts_in_cur_img` and it is then broadcasted to divide each column of `src_pts_in_cur_img`

Reply:
yes, but you are dividing `z / z`, which will make all `z` values to be `1.0`.

The correct transform is what I wrote in `uncompute_pcd`, which is only `[:, 0] / [:, 2]` and `[:, 1] / [:, 2]` need to be done.

Reply:
That was intended - wanted to keep the homogenous image coordinates in the format `(x,y,1)`.